server/src/main/java/org/elasticsearch/search/aggregations/timeseries/TimeSeriesIndexSearcher.java

```diff
@@ -0,0 +1,133 @@
+import org.apache.lucene.search.Weight;
+import org.apache.lucene.util.Bits;
+import org.apache.lucene.util.BytesRef;
+import org.apache.lucene.util.PriorityQueue;
+import org.elasticsearch.cluster.metadata.DataStream;
+import org.elasticsearch.index.mapper.TimeSeriesIdFieldMapper;
+import org.elasticsearch.search.aggregations.BucketCollector;
+import org.elasticsearch.search.aggregations.LeafBucketCollector;
+
+import java.io.IOException;
+
+/**
+ * An IndexSearcher wrapper that executes the searches in time-series indices by traversing them by tsid and timestamp
+ * TODO: Convert it to use index sort instead of hard-coded tsid and timestamp values
+ */
+public class TimeSeriesIndexSearcher {
+
+    private final IndexSearcher searcher;
+
+    public TimeSeriesIndexSearcher(IndexSearcher searcher) {
+        this.searcher = searcher;
+    }
+
+    public void search(Query query, BucketCollector bucketCollector) throws IOException {
+        query = searcher.rewrite(query);
+        Weight weight = searcher.createWeight(query, bucketCollector.scoreMode(), 1);
+        PriorityQueue<LeafWalker> queue = new PriorityQueue<>(searcher.getIndexReader().leaves().size()) {
+            @Override
+            protected boolean lessThan(LeafWalker a, LeafWalker b) {
+                assert a.tsid != null && a.timestamp != null && b.tsid != null && b.timestamp != null;
+                int res = a.tsid.compareTo(b.tsid);
+                if (res == 0) {
+                    return a.timestamp < b.timestamp;
+                } else {
+                    return res < 0;
+                }
+            }
+        };
+        for (LeafReaderContext leaf : searcher.getIndexReader().leaves()) {
+            LeafBucketCollector leafCollector = bucketCollector.getLeafCollector(leaf);
+            Scorer scorer = weight.scorer(leaf);
+            if (scorer != null) {
+                LeafWalker walker = new LeafWalker(leaf, scorer, leafCollector);
+                if (walker.next()) {
+                    queue.add(walker);
+                }
+            }
+        }
+        while (queue.top() != null) {
+            LeafWalker walker = queue.top();
+            walker.collectCurrent();
+            if (walker.next()) {
+                queue.updateTop();
+            } else {
+                queue.pop();
+            }
+        }
+    }
+
+    private static class LeafWalker {
+        private final LeafCollector collector;
+        private final Bits liveDocs;
+        private final DocIdSetIterator iterator;
+        private final SortedSetDocValues tsids;
+        private final SortedNumericDocValues timestamps;
+        int docBase;
+        int docId;
+        BytesRef tsid;
+        Long timestamp;
+
+        String id;
+
+        LeafWalker(LeafReaderContext context, Scorer scorer, LeafCollector collector) throws IOException {
+            id = context.toString().replaceFirst(".*:c", "c");
+            id = id.substring(0, id.indexOf(':'));
```

Comment:
These are probably worth a comment on what they do. Maybe even an assertion that the result matches a pattern you expect it to match.

Reply:
Debug code leftovers.